File: webapp/api/services/storage-service.ts
import fs from 'fs'
import path from 'path'
import OSS from 'ali-oss'
import { config, isOssEnabled } from '../config.js'

// #region debug-point A:storage-reporter
const reportDebug = (
  msg: string,
  data: Record<string, unknown> = {},
) => {
  void fetch('http://127.0.0.1:7777/event', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      sessionId: 'upload-image-500',
      runId: 'post-fix',
      hypothesisId: 'A',
      location: 'api/services/storage-service.ts',
      msg: `[DEBUG] ${msg}`,
      data,
      ts: Date.now(),
    }),
  }).catch(() => {})
}
// #endregion

let ossClient: OSS | null = null

if (isOssEnabled) {
  ossClient = new OSS({
    region: config.ossRegion,
    accessKeyId: config.ossAccessKeyId,
    accessKeySecret: config.ossAccessKeySecret,
    bucket: config.ossBucket,
    endpoint: config.ossEndpoint,
  })
}

export function getStorageMode(): 'oss' | 'local' {
  return isOssEnabled ? 'oss' : 'local'
}

export async function storeImage(file: Express.Multer.File) {
  const imageId = crypto.randomUUID()
  const extension = path.extname(file.originalname) || '.jpg'
  const objectKey = `ingredients/${new Date().toISOString().slice(0, 10)}/${imageId}${extension}`
  // #region debug-point A:store-entry
  reportDebug('storeImage entered', {
    fileName: file.originalname,
    mimeType: file.mimetype,
    size: file.size,
    objectKey,
    storageMode: getStorageMode(),
  })
  // #endregion

  if (ossClient) {
    // #region debug-point A:oss-put-start
    reportDebug('oss put starting', {
      objectKey,
      endpoint: config.ossEndpoint,
      bucket: config.ossBucket,
      region: config.ossRegion,
    })
    // #endregion
    const result = await ossClient.put(objectKey, file.buffer, {
      headers: {
        'Content-Type': file.mimetype,
      },
    })
    // #region debug-point A:oss-put-success
    reportDebug('oss put succeeded', {
      objectKey,
      url: result.url,
      name: result.name,
    })
    // #endregion

    return {
      imageId,
      objectKey,
      imageUrl: result.url,
      fileName: file.originalname,
      contentType: file.mimetype,
    }
  }

  fs.mkdirSync(config.uploadDir, { recursive: true })

  const filePath = path.join(config.uploadDir, `${imageId}${extension}`)
  fs.writeFileSync(filePath, file.buffer)
  // #region debug-point A:local-write-success
  reportDebug('local file write succeeded', {
    filePath,
    objectKey: path.basename(filePath),
  })
  // #endregion

  return {
    imageId,
    objectKey: path.basename(filePath),
    imageUrl: `/uploads/${path.basename(filePath)}`,
    fileName: file.originalname,
    contentType: file.mimetype,
  }
}

File: webapp/backend/services/storage_service.py
import uuid
from pathlib import Path

from ..config import IS_OSS_ENABLED, OSS_ACCESS_KEY_ID, OSS_ACCESS_KEY_SECRET, OSS_BUCKET, OSS_ENDPOINT

# 本地文件存储目录
LOCAL_UPLOAD_DIR = Path(__file__).resolve().parent.parent.parent / "storage" / "uploads"


def get_storage_mode() -> str:
    # 返回当前存储模式：oss 或 local
    return "oss" if IS_OSS_ENABLED else "local"


async def store_image(file_bytes: bytes, filename: str, content_type: str) -> dict:
    # 存储上传的图片，支持本地存储和阿里云 OSS 两种方式
    image_id = str(uuid.uuid4())
    ext = Path(filename).suffix or ".jpg"
    object_key = f"chat/{image_id}{ext}"

    if IS_OSS_ENABLED:
        # 上传到阿里云 OSS
        import oss2
        auth = oss2.Auth(OSS_ACCESS_KEY_ID, OSS_ACCESS_KEY_SECRET)
        bucket = oss2.Bucket(auth, OSS_ENDPOINT, OSS_BUCKET)
        result = bucket.put_object(object_key, file_bytes, headers={"Content-Type": content_type})
        image_url = result.url if hasattr(result, "url") else f"https://{OSS_BUCKET}.{OSS_ENDPOINT}/{object_key}"
    else:
        # 保存到本地文件系统
        LOCAL_UPLOAD_DIR.mkdir(parents=True, exist_ok=True)
        file_path = LOCAL_UPLOAD_DIR / f"{image_id}{ext}"
        file_path.write_bytes(file_bytes)
        image_url = f"/uploads/{image_id}{ext}"

    return {
        "imageId": image_id,
        "imageUrl": image_url,
        "fileName": filename,
        "contentType": content_type,
    }
